src/vmm/src/guest_config/x86_64/mod.rs

```diff
@@ -1,17 +1,359 @@
 // Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-use std::collections::HashMap;
+/// Module with CPU templates for x86_64
+pub mod static_cpu_templates;
+use std::collections::{HashMap, HashSet};
 
+use arch::x86_64::msr::create_boot_msr_entries;
+use kvm_bindings::kvm_msr_entry;
+use static_cpu_templates::*;
+
+use super::cpuid::{CpuidKey, RawCpuid};
+use super::templates::x86_64::{CpuidRegister, X86_64CpuTemplate};
 use crate::guest_config::cpuid::Cpuid;
+use crate::vmm_config::machine_config::CpuTemplateType;
+use crate::vstate::vcpu::msr_entries_to_save;
+use crate::vstate::vm::Vm;
+
+/// Errors thrown while configuring templates.
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+pub enum Error {
+    /// Failure in processing the CPUID in template for x86_64 CPU configuration.
+    #[error(
+        "Template changes a CPUID entry not supported by the host: Leaf: {0:0x}, Subleaf: {1:0x}"
+    )]
+    CpuidFeatureNotSupported(u32, u32),
+    /// Failure in processing the MSRs in template for x86_64 CPU configuration.
+    #[error("Template changes an MSR entry not supported by the host: Register Address: {0:0x}")]
+    MsrNotSupported(u32),
+    /// Can not join 2 cpuids.
+    #[error("Can not join 2 cpuids: {0}")]
+    CpuidJoin(super::cpuid::CpuidJoinError),
+    /// Can create cpuid from raw.
+    #[error("Can create cpuid from raw: {0}")]
+    CpuidFromRaw(super::cpuid::CpuidTryFromRawCpuid),
+}
 
 /// CPU configuration for x86_64 CPUs
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Debug, Clone, PartialEq)]
 pub struct X86_64CpuConfiguration {
     /// CPUID configuration
     pub cpuid: Cpuid,
     /// Register values as a key pair for model specific registers
     /// Key: MSR address
     /// Value: MSR value
     pub msrs: HashMap<u32, u64>,
+
+    /// Set of supported MSRs
+    pub msr_list: HashSet<u32>,
+
+    /// Architectural MSPs required for boot
+    pub msr_boot_entries: Vec<kvm_msr_entry>,
```

Comment:
we need this field, because some static templates need to add msrs to it.

Reply:
We don't need to store `supported_msrs` and `msr_boot_entries` in `CpuConfiguration`.
- `supported_msrs` is also stored in `msr_list` of `KvmVcpu` and the `msr_list` of `KvmVcpu` can be extended in `KvmVcpu::configure()` by using `msrs` of `CpuConfiguration`
- `msr_boot_entries` should be placed in `KvmVcpu::configure()` because it should be applied regardless of static or custom CPU templates.

Reply:
So in `CpuConfiguration`, we should only construct `msrs` that will be joined into `msr_boot_entries` and be added into the `msr_list` of `KvmVcpu` in `KvmVcpu::configure()`.

Reply:
We can handle this in Nikita's PR #3596. Closing this here.